File: Sample/GameRanking/Server/release/Script/PyScript/Action/action1001.py
"""1001_排名列表接口"""
import ReferenceLib
from action import *
from ZyGames.Framework.Common import *
from ZyGames.Framework.Cache.Generic import *
from ZyGames.Framework.Game.Service import *
from GameServer.Model import *

class UrlParam(HttpParam):
    def __init__(self):
        HttpParam.__init__(self)
        self.PageIndex = 0
        self.PageSize = 0


class ActionResult(DataResult):
    def __init__(self):
        DataResult.__init__(self)
        self.PageCount = 0
        self.List = []


def getUrlElement(httpGet, parent):
    urlParam = UrlParam()
    if httpGet.Contains("PageIndex")\
    and httpGet.Contains("PageSize"):
        urlParam.PageIndex = httpGet.GetIntValue("PageIndex")
        urlParam.PageSize = httpGet.GetIntValue("PageSize")
    else:
        urlParam.Result = False

    return urlParam
   
def compareTo(x, y):
    return y.Score - x.Score

def takeAction(urlParam, parent):
 
    actionResult = ActionResult()
    cache = ShareCacheStruct[UserRanking]()
    actionResult.List = cache.FindAll()
    actionResult.List = MathUtils.QuickSort[UserRanking](actionResult.List, compareTo)
    pagingResult = MathUtils.GetPaging[UserRanking](actionResult.List, urlParam.PageIndex, urlParam.PageSize)
    if pagingResult :
        actionResult.List = pagingResult[0]
        actionResult.PageCount = pagingResult[1]

    return actionResult

def buildPacket(writer, urlParam, actionResult):
    writer.PushIntoStack(actionResult.PageCount)
    writer.PushIntoStack(len(actionResult.List))
    for info in actionResult.List:
        dsItem = DataStruct()
        dsItem.PushIntoStack(info.UserName)
        dsItem.PushIntoStack(info.Score)
        writer.PushIntoStack(dsItem)


    return True

File: Source/Middleware/GameServer/Script/CsScript/Action/Action1001.cs
/****************************************************************************
Copyright (c) 2013-2015 scutgame.com

http://www.scutgame.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System;
using System.Collections.Generic;
using GameServer.Model;
using ZyGames.Framework.Cache.Generic;
using ZyGames.Framework.Common;
using ZyGames.Framework.Game.Contract;
using ZyGames.Framework.Game.Service;

namespace GameServer.CsScript.Action
{
    public class Action1001 : BaseStruct
    {
        private int PageIndex;
        private int PageSize;
        private int PageCount;
        private List<UserRanking> rankingList;


        public Action1001(HttpGet httpGet)
            : base(1001, httpGet)
        {

        }

        public override void BuildPacket()
        {
            this.PushIntoStack(PageCount);
            this.PushIntoStack(rankingList.Count);
            foreach (var item in rankingList)
            {
                DataStruct dsItem = new DataStruct();
                dsItem.PushIntoStack(item.UserName);
                dsItem.PushIntoStack(item.Score); 
                //Console.WriteLine("Num count:{0}", item.Items.Count);
                this.PushIntoStack(dsItem);
            }

        }

        public override bool GetUrlElement()
        {
            if (httpGet.GetInt("PageIndex", ref PageIndex)
                 && httpGet.GetInt("PageSize", ref PageSize))
            {
                return true;
            }
            return false;
        }

        public override bool TakeAction()
        {
            var cache = new ShareCacheStruct<UserRanking>();
            rankingList = cache.FindAll(false);
            rankingList = MathUtils.QuickSort<UserRanking>(rankingList, compareTo);
            rankingList = rankingList.GetPaging(PageIndex, PageSize, out PageCount);
            return true;
        }

        private int compareTo(UserRanking x, UserRanking y)
        {
            int result = y.Score - x.Score;
            if (result == 0)
            {
                result = y.UserID - x.UserID;
            }
            return result;
        }
    }
}
